Stepper: fix the signup completion event.

southp opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to Automattic/growth-foundations#70, and is based on https://github.com/Automattic/wp-calypso/pull/78323.

## Proposed Changes

This PR improves the signup completion event issues discussed in the linked issue above. Namely:

* There are two `calypso_signup_completion` fired, one from the classic signup framework, `/start/account`, and one from the Stepper itself.
* The `calyspo_signup_completion` event from Stepper is not complete. 
* It's not correct to fire the signup completion event on Launchpad, since it's not really part of the signup flow. The users who haven't launched their sites can always see it the 1st thing clicking on the "home" button.

Since [the parent PR](https://github.com/Automattic/wp-calypso/pull/78323) uses the Newsletter flow as the updated case, the fix for the doubled `calypso_signup_completion` only works for the Newsletter flow for now. The other fixes apply to all the tailored flows.

Note that, in case of completing the signup completion event, this PR only fills the purchase info that's most relevant to the linked Growth Foundation issue. Filling in the other fields is intentionally left out as follow-ups.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

#### The doubled signup completion event

* Follow the instruction in [the analytics lib README](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/analytics/README.md#L13) and enable the console debugging message, i.e. calling `localStorage.setItem('debug', 'calypso:analytics*');`. Since there will be redirection, make sure to enable "preserve log" for your browser console.
* Go to the Newsletter flow, `/setup/newsletter`
* Complete the user step, and make sure that there is no `calypso_signup_complete` event.
* Cherry-pick some other flows and do the same and note that there is still `calypso_signup_complete` event fired as expected, since they are not updated yet.

#### The more completed signup completion event and its timing

In general:

1. Go through the Newsletter flow til the end of the processing step
2. Make sure `calypso_signup_complete` event is fired with expected value. 
3. Confirm Launchpad no longer fires the signup completion event.

Note that the console message tends to be truncated in this case, so I'd reommend to observe it from the Network tab instead, e.g.
<img width="655" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1842898/24b32e62-be79-45c8-b1cb-5dc9d53524eb">

Combinations worth trying and the values to pay attention to:

* A free domain + Free plan -> `hasCartItem`: false, no `plan_product_slug` and no `domain_product_slug` fields.
* A paid domain + Free plan (since it is currently allowed on this flow, it still worth examining) -> `hasCartItem`: true, has `domain_product_slug`, no `plan_product_slug`
* A paid domain + a paid plan -> `hasCartItem`: true, `plan_product_slug` and `domain_product_slug` should be populated as expected. 

Note that it's not possible to test out the domain mapping and the domain transfer case yet. The best for now is to compare it to the original code: https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/main.jsx#L501.

#### Misc.

Cherry-pick a few other tailored flows, e.g. link in bio, and confirm the section of "The more completed signup completion event and its timing" also works as expected.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

southp committed:
Fire the signup completion event upon all the processing actions being
completed in the processing step.

southp committed:
Fill in the missing props of the signup completion event as much as
possible

southp committed:
Do not fire the signup completion event in the launchpad, since it
shouldn't really act as part of the signup flow.

southp committed:
remove `recordSignupComplete` from Launchpad

southp committed:
Fix product slug derivation logics and remove `now`

southp commented on client/landing/stepper/hooks/use-record-signup-complete.ts:
```diff
@@ -1,37 +1,66 @@
-				isMapping: false,
-				isTransfer: false,
-			},
-			true
```
This `now` parameter is removed for two reasons:

1. From what I recall, it's not recommended unless there is a very good reason. The Tracks event queue is there to guarantee the order of the Tracks event is intact. By introducing `now` to force it to fire immediately, there can be edged chances that the order being ruined.
2. [The main onboarding flow](https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/main.jsx#L501) has been working this way for ages :)

southp committed:
A vital typo: domainCarItem => domainCartItem

southp committed:
Fix typing errors.

skim1220 commented:
👋 @southp (cc: @jdc91), I finished reviewing the flow in this PR.

## Observations per step in the flow

### 1. Landing page 

I noticed that signup start events (`calypso_signup_start`, `calypso_singup_step_start`) are fired on the landing page in this test flow (see Row 10-11 in [this Google Sheet](https://docs.google.com/spreadsheets/d/1hFq9nXeKpnfcZOwLE-kgiiwf7DEk80KKKUV_a9ZdRh0/edit?usp=sharing)). 

⚠️ This should NOT happen. The landing page is not a part of signup flow, so these 2 signup start events should not fire on this page but on “Let’s get started” page.

### 2. Let’s get started page

I was able to see that it starts account flow (not onboarding flow) and the page path began with `/start/...`.

In the `onboarding` flow, `calypso_signup_start` event logs `ref` property on this page, which tells us the source landing page. However, in the `account` flow,  `calypso_signup_start` event does _**not**_ record `ref` property. 

Is it possible to record the source landing page as `ref` property of `calypso_signup_start` event on this page in the `account` flow, just like we do in the `onboarding flow? This is not a blocker, but it will be helpful when we audit the data afterwards.  

### 3. Newsletter setup page

I was able to see that the path changed from `/start/...`  to `/setup/...` on this page. 
Since this is not a part of onboarding flow, I only checked if flow name is consistent with previous step, which is the case. 

### 4. Domain Selection 

“_Use a domain I own_” option is not available, so I wasn’t able to test it in this flow. 

I was able to see the following Tracks events fired on this page, and their properties are consistent: 
• `calypso_signup_actions_submit_step` (`flow = newsletter`) 
• `calypso_signup_step_start` (`flow = newsletter`,  `step = domains`)
• `calypso_domain_search_pageview` (`section = signup`) 
• `calypso_page_view` (`path = /setup/newsletter/domains`) 

### 5. Plan Selection 

I was able to see the following Tracks events fired on this page, and their properties are consistent: 
• `calypso_signup_skip_step` (`flow = newsletter`, step = domains`, `section = signup`) 
• `calypso_domain_search_add_button_click` (`section = signup`)  
• `calypso_domain_search_submit_step` (`section = signup`, `type = domain_reg or  wpcom_subdomain or dotblog_subdomain`)
• `calypso_signup_actions_submit_step` (`flow = newsletter`, `step = domains`)
• `calypso_signup_step_start` (`flow = newsletter`, `step = plans`) 
• `calypso_page_view` (`path = /setup/newsletter/plans`) 

⚠️ Note that I was able to proceed with Free Plan even when I selected Custom Domain. **_This should not happen_**. In this case, we should show a modal on Plan Selection page explaining this. 

⚠️ One thing I’d like to double check: I noticed that we have a toggle between “Monthly” and “Yearly” ([screenshot](https://d.pr/i/cvtL35)). This is still the case in the current onboarding flow. However, for our project, we will remove “Monthly” option on this page, and instead have a toggle between “1-year” and “2-year”. Is this understanding correct? (cc: @demetd , @isatuncman-auto)

### 6-1 Add your subscriber page

Whenever a user does not have any paid product in their cart, they are directed to this page, and this is when `calypso_signup_complete` event is fired. This page seems equivalent to “What are your goals?” page in the current onboarding flow. 

⚠️ Note that **`calypso_signup_complete` event properties do not align with how it works in the current onboarding flow**.  See the separate section below. 

### 6-2 Checkout Page 

Whenever a user has any paid product in the cart, they are directed to the checkout page, and this is when `calypso_signup_complete` event is fired. 

I noticed that `calypso_checkout_sidebar_upsell_click` event records `switching_from` and `switching_to` properties. We can easily parse this to understand shifting between different plans, which sounds helpful. Just an idea (but definitely outside of the current project scope), is it possible to record `switching_from` and `switching_to` properties with `calypso_checkout_composite_plan_length_change` event as well? This event fires when a user switches between different plans in the dropdown menu. 

⚠️ I didn’t proceed with the checkout, but we should make sure that `calypso_signup_complete` event is not firing again after the checkout page.  

⚠️ `calypso_signup_complete` event properties do not align with how it works in the current onboarding flow.  See the separate section below. 


## `calypso_signup_complete` event properties

I see some misalignment between how the flow in this PR logs `calypso_signup_complete` properties vs. how the current onboarding does it. Please see below where they do not align. 

What happens in the flow in this PR:
![Screenshot on 2023-06-22 at 13_27_26](https://github.com/Automattic/wp-calypso/assets/80088175/27b18b21-5edf-42e0-bc8e-7ccfef75c78e)
([image link](https://d.pr/i/HbYf9p))

How they should be: 
![Screenshot on 2023-06-22 at 13_31_13](https://github.com/Automattic/wp-calypso/assets/80088175/29d4ad4c-f5de-46bc-aee1-094f539b7cc7)
([image link](https://d.pr/i/yCHBIi))

FYI, `signup_domain_origin` is a new property we merged yesterday ([PR](https://github.com/Automattic/wp-calypso/pull/78085)), and my review results are [here](https://github.com/Automattic/growth-foundations/issues/71#issuecomment-1602305886). 

These event properties of `calypso_signup_complete` are crucial in our signup cohort analysis. Can we make sure that they are working correctly? 

Let me know if I anything above is not clear or if I can help with anything else 🙏

demetd commented:
This audit is amazing @skim1220 💯 cc @elekanna 

> ⚠️ One thing I’d like to double check: I noticed that we have a toggle between “Monthly” and “Yearly” ([screenshot](https://d.pr/i/cvtL35)). This is still the case in the current onboarding flow. However, for our project, we will remove “Monthly” option on this page, and instead have a toggle between “1-year” and “2-year”. Is this understanding correct? (cc: @demetd , @isatuncman-auto)

This is correct!

jdc91 commented:
@skim1220 Thanks for the deep analysis, as discussed elsewhere ( p1687445830131569/1687350686.732589-slack-C059R8S4ALT ) I will be focussing in on the feedback related to `calypso_signup_complete`. It's good if we can carry the rest of the feedback to a more globally trackable location like an issue.

Also as noted in the PR description above:

> Note that it's not possible to test out the domain mapping and the domain transfer case yet. The best for now is to compare it to the original code: https://github.com/Automattic/wp-calypso/blob/trunk/client/signup/main.jsx#L501.

So we will have to monitor and fix this later. This is another possible issue that we might have to create and track but I will wait on your feedback as to how to manage the project.


### Fixes Planned

I will be focussed on the red outlined issues highlighted in the table your provided, included below

![image](https://github.com/Automattic/wp-calypso/assets/3422709/5ac8f7e1-5a86-49bc-8a38-ef7bbe338af8)

jdc91 commented:
## Ongoing Self notes 
On the onboarding flow. We seem to be setting the `domainCartItem` as undefined when it is free.

https://github.com/Automattic/wp-calypso/blob/6f87983e61d1dd9b5f4dd8a057ffcb41ef426b72/client/landing/stepper/declarative-flow/internals/steps-repository/choose-a-domain/index.tsx#L77-L97

southp committed:
Add a `toStepper` mechanism to avoid the signup completion event from
firing. It is specifically for Stepper using /start/account to create an
account.

ddc22 committed:
Fix type error where the type should be string

ddc22 reviewed (approved):
LGTM, Code looks good.
I have also done some basic tests with the `newsletter` and `link-in-bio` flow and they works as expected with the caveats above highlighted by @skim1220 still remaining to be resolved.
 Rest of the issues will be picked up in subsequent issues Automattic/martech#1878
 🚢
